core/src/main/java/io/grpc/transport/MessageDeframer.java

```diff
@@ -124,27 +148,131 @@ public MessageDeframer(Listener listener) {
   public MessageDeframer(Listener listener, Compression compression) {
     this.listener = Preconditions.checkNotNull(listener, "sink");
     this.compression = Preconditions.checkNotNull(compression, "compression");
+    queueRef.set(new ArrayDeque<Frame>(4));
   }
 
   /**
    * Requests up to the given number of messages from the call to be delivered to
-   * {@link Listener#messageRead(InputStream)}. No additional messages will be delivered.
+   * {@link Listener#messagesAvailable}. No additional messages will be delivered.
+   * Only called by the application thread.
    *
    * <p>If {@link #close()} has been called, this method will have no effect.
    *
    * @param numMessages the requested number of messages to be delivered to the listener.
    */
-  public void request(int numMessages) {
+  void request(int numMessages) {
     Preconditions.checkArgument(numMessages > 0, "numMessages must be > 0");
     if (isClosed()) {
       return;
     }
-    pendingDeliveries += numMessages;
-    deliver();
+    // Make the number of requested messages visible to the transport thread.
+    transportCount.addAndGet(numMessages);
+    requestedMessageCount += numMessages;
+    if (messagesAvailable && !appInRequest) {
+      appInRequest = true;
+      try {
+        listener.messagesAvailable(this);
+      } finally {
+        appInRequest = false;
+      }
+    }
+  }
+
+  /**
+   * Called in the application thread to consume the available messages provided by the transport.
+   * @param consumer which receives all available messages.
+   */
+  @Override
+  public int drainTo(StreamListener.MessageConsumer consumer) {
+    int count = 0;
+    do {
+      if (appQueue == null) {
+        appQueue = new ArrayDeque<Frame>();
```

Comment:
nit: Should probably create the new deque the same way we do in the constructor.  Maybe a utility method?